File: leetcode/_20.cpp
#include <stack>
#include <unordered_map>
class Solution
{
public:
    bool isValid(std::string s)
    {
        std::stack<char> st;
        for (char c : s)
        {
            // 如果是左括号
            if (c == '(' || c == '{' || c == '[')
                st.push(c);
            else
            {
                if (st.empty())
                    return false;

                if (st.top() == '(')
                {
                    if (c == ')')
                        st.pop();
                    else
                        return false;
                }
                else if (st.top() == '{')
                {
                    if (c == '}')
                        st.pop();
                    else
                        return false;
                }
                else if (st.top() == '[')
                {
                    if (c == ']')
                        st.pop();
                    else
                        return false;
                }
            }
        }

        return st.empty();
    }
};

File: leetcode/_20.py
class Solution:
    def isVaild(self, s:str) ->bool:
        stack = []
        for c in s:
            if c in "([{":
                stack.append(c)
            else:
                if not stack:
                    return False
                if stack[-1] == "(" and c == ")":
                    stack.pop()
                elif stack[-1] == "[" and c == "]":
                    stack.pop()
                elif stack[-1] == "{" and c == "}":
                    stack.pop()
                else:
                    return False
            
        return not stack